### ISSUE TITLE
Confusing default `stride_3D` parameter for architecture `Modified3DUNet`

### ISSUE THREAD

**Issue by jcohenadad** [ivadomed/ivadomed#1248]
### Motivation for the feature

The documentation for the `Modified3DUNet` architecture shows:

![image](https://user-images.githubusercontent.com/2482071/208773727-c3a9d974-497f-4294-b39d-75dca6e0d4aa.png)

My understanding of a stride, confirmed with the description, is the number of skipped voxels during convolution. However, the default stride is the size of the default length, ie: `[128, 128, 16]`, which essentially means that there is only one pass of the convolution?

**Comment by jcohenadad** [comment]
However, I am still confused by the `stride_3D` value. When specifying this config:
```json
        "length_3D": [32, 64, 64],
        "stride_3D": [1, 1, 2],
        "CenterCrop": {"size": [32, 64, 128]},
```
Does it mean that there will be two patches per volumes? (based on the visual output, it seems to be the case). So, in other words, if CenterCrop_x = length_3D_x, does it mean that there is no striding done along x?

**Comment by naga-karthik** [comment]
> a stride typically refers to the convolution of the kernel on the patch

Firstly, to clarify this a bit more, stride (in typical DL terms) refers to the number of pixels/voxels the convolutional kernel is expected to shift in order to perform the convolution on the image patch. The stride is typically a small number (1 or 2, in most cases) which tells that the kernel is shifted by 1 pixel/voxel in order to perform the convolution. 

However, this is NOT the same as what ivadomed describes in `stride_3D`, which is a slight abuse of the term and hence causes confusion. 

First, the original stride (that is, how everyone in DL understands it, is fixed for 3D UNet model can be seen in any convolution layer, for e.g. [here](https://github.com/ivadomed/ivadomed/blob/master/ivadomed/models.py#L799)). What `stride_3D` refers to is how far or how close you want the patches to be extracted from the center cropped volume (note that this has nothing to do with the shifting of the convolutional kernel - this is fixed as shown above). Taking the example you have given: in cropped volume of size `[32, 64, 128]`, one patch is extracted at the location `[32, 64, 64]`. For the next patch, since `stride_3D` is `[1, 1, 2]`, the next patch is extracted at the location `[32, 64, 66]`, then at `[32, 64, 68]` and this goes on until the entire patch of size `[32, 64, 128]` is traversed. So, this means that there are 64 such patches that are extracted if `stride_3D` is `[1, 1, 2]`. One can observe that _based on this definition of `stride_3D`_, if small values are chosen then there will be a lot of overlap between the patches that are extracted. 

> Does it mean that there will be two patches per volumes? 

If 2 volumes per center-cropped patch are desired, then stride_3D should be `[32, 64, 64]`. That is, one patch at this location and the next patch is extracted after moving 64 voxels in the 3rd dimension. 

> So, in other words, if CenterCrop_x = length_3D_x, does it mean that there is no striding done along x?

Yes, because the length of the patch is precisely how much of the input has been cropped thus leaving no space for more patches to be extracted from this dimension. 

Let me know if this explanation makes sense! It is understandable that this is confusing (not sure why the names `length_3D` and `stride_3D` were chosen to begin with), especially when _stride_ has a different meaning universally.

**Comment by jcohenadad** [comment]
Thank you for the clarification @naga-karthik . My problem is that, when setting the Stride_3D parameter to what it _should_ be, I am running into https://github.com/ivadomed/ivadomed/issues/1245 and/or https://github.com/ivadomed/ivadomed/issues/1022#issuecomment-1371640731. The latter being particularly problematic, because it _forces_ me to use a larger input image size than what would be reasonably expected (ie: if dim is 128 and kernel is 64, I expect that a "stride" of 64 should be sufficient, but according to the [definition](https://github.com/ivadomed/ivadomed/issues/1022) it is not).

Moreover, me using `Stride_3D` of 1 with "CenterCrop_x = length_3D_x" produces a SINGLE patch (when considering one dimension, for clarity of the demonstration). That is also quite confusing and we should do something about it (eg: output a warning or something). 

I'll follow-up with a calculation of patch generation with a current experiment, just to make sure we are all on the same page.

**Comment by jcohenadad** [comment]
Here's the calculation. 

config:
```json
length_3D: [32, 64, 64]
stride_3D: [1, 1, 2]
CenterCrop: {'size': [32, 64, 128]}
```

Data: Based on the file `bids_dataframe.csv` (it would be more convenient to have the info spitted out by the logger-- related to https://github.com/ivadomed/ivadomed/issues/1102), there are 178 paired with a label. 

Ivadomed output:
~~~
2023-01-05 13:34:38.166 | INFO     | ivadomed.loader.loader:load_dataset:118 - Loaded 1815 volumes of shape [32, 64, 64] for the validation set.
2023-01-05 13:34:58.191 | INFO     | ivadomed.loader.loader:load_dataset:118 - Loaded 3498 volumes of shape [32, 64, 64] for the training set.
~~~

**Comment by naga-karthik** [comment]
I think I can understand what's causing this (though I could not get the exact number of volumes as shown - that depends on the number of subjects used for training and validation). The magic is happening in [these lines](https://github.com/ivadomed/ivadomed/blob/master/ivadomed/loader/mri3d_subvolume_segmentation_dataset.py#L144-L154) inside the class `MRI3DSubVolumeSegmentationDataset` (same shown below). 

```
for x in range(0, (shape[0] - self.length[0]) + 1, self.stride[0]):
    for y in range(0, (shape[1] - self.length[1]) + 1, self.stride[1]):
        for z in range(0, (shape[2] - self.length[2]) + 1, self.stride[2]):
            self.indexes.append({
                'x_min': x,
                'x_max': x + self.length[0],
                'y_min': y,
                'y_max': y + self.length[1],
                'z_min': z,
                'z_max': z + self.length[2],
                'handler_index': I})
```

So, what I explained in my comment above remains as it is - the stride_3D is essentially the "step" the we use to skip indices in the standard `for` loop. Now, with your parameters, we have the following:

```
shape: [32, 64, 128]    # this is the shape printed in the code after center cropping
length: [32, 64, 64]
stride: [1, 1, 2]

for x in range(0, 1, 1):
    for y in range(0, 1, 1):
        for z in range(0, 65, 2):
            self.indexes.append({
                'x_min': x,
                'x_max': x + 32,
                'y_min': y,
                'y_max': y + 64,
                'z_min': z,
                'z_max': z + 64,
                'handler_index': I})
```
so, for each center-cropped volume of size `[32, 64, 128]`, **33 sub-volumes are extracted** (not 32). Now, if the number of subjects for validation are `n` then, `n x 33` should give you the total number of volumes in the validation set, which is 1815.

**Comment by jcohenadad** [comment]
Thank you for the digging, @naga-karthik, but I was not too worried about the non exact match (further digging would explain it, but that's not the point). The two main worries where we need to act are
1. The first paragraph in https://github.com/ivadomed/ivadomed/issues/1248#issuecomment-1372748811
2. The confusion about the Stride_3D naming (more of a UX issue)

**Comment by naga-karthik** [comment]
Ah okay! Considering this:

> (ie: if dim is 128 and kernel is 64, I expect that a "stride" of 64 should be sufficient, but according to the https://github.com/ivadomed/ivadomed/issues/1022 it is not).

Would you providing the exact configuration of length, stride and center-crop that's not working? I'm asking because every number is a multiple of 16 so I was wondering which configuration is raising an error

As for this: 

> The confusion about the Stride_3D naming (more of a UX issue)

`step_size_3D` should be a good alternative

**Comment by jcohenadad** [comment]
I did another experiment, to "fix" my wrong patch coverage along the 3rd dimension:
```json
"train_fraction": 0.6,
        "test_fraction": 0.1
        "length_3D": [32, 64, 64]
        "stride_3D": [1, 1, 64]. <---- that's the change! (previously it was 1,1,2
        "CenterCrop": {'size': [32, 64, 128]}
```

Data: Based on the file bids_dataframe.csv (it would be more convenient to have the info spitted out by the logger-- related to https://github.com/ivadomed/ivadomed/issues/1102), there are 178 subjects paired with a label (same as before).

Ivadomed output:
```console
2023-01-05 17:49:03.610 | INFO     | ivadomed.loader.loader:load_dataset:118 - Loaded 110 volumes of shape [32, 64, 64] for the validation set.
2023-01-05 17:49:23.592 | INFO     | ivadomed.loader.loader:load_dataset:118 - Loaded 212 volumes of shape [32, 64, 64] for the training set.
```

So, 0.6*178 = 106.8 subjects for training. Considering a stride of 64 along Z, it takes 2 steps to go from 64 to 128, so 2 patches per subjects, which gives 106.8*2 = 213 patches, which is consistent with the "212" patches mentioned in the ivadomed output. So, the numbers again make sense. And I'm happy that issue number one listed in https://github.com/ivadomed/ivadomed/issues/1248#issuecomment-1372887236 is now solved. However we need to work on the UX.

**Comment by jcohenadad** [comment]
> step_size_3D should be a good alternative

Yes, although I'd like it if we could finally unify our 2D and 3D fields/codebases

**Comment by jcohenadad** [comment]
> EDIT: changed 0.6178 = 106.8 originally to 0.6*178 = 106.8

yes, excellent point

**Comment by jcohenadad** [comment]
> Would you mind providing the exact configuration of length, stride and center-crop that's not working? I'm asking because every number is a multiple of 16 so I was wondering which configuration is raising an error

Another example here https://github.com/ivadomed/ivadomed/issues/1022#issuecomment-1371640731, with no multiple of 16-- it should fail, and it does not fail

**Comment by naga-karthik** [comment]
Ah yes, I just checked that, seems like the reason for setting the "multiple of 16" condition is indeed mysterious.

**Comment by naga-karthik** [comment]
Based on a brief discussion about unifying the current "stride" key across 2D and 3D models, here's what was decided: 
At this stage, the unification of 2D and 3D for this key requires major refactoring of the codebase, hence we would like to take small steps and change the (currently confusing) key name from `stride_3D` to `patch_step_size_3D` for the 3D models and `patch_step_size_2D` for the 2D models for now. Future modifications to the updated key will only require it to be changed to `patch_step_size`, thus removing the suffixes for both 2D and 3D.  
(tagging @mariehbourget for confirmation - please feel free to edit if you want to add anything more. Thanks!)

**Comment by mariehbourget** [comment]
> Based on a brief discussion about unifying the current "stride" key across 2D and 3D models, here's what was decided:
> At this stage, the unification of 2D and 3D for this key requires major refactoring of the codebase, hence we would like to take small steps and change the (currently confusing) key name from `stride_3D` to `patch_step_size_3D` for the 3D models and `patch_step_size_2D` for the 2D models for now. Future modifications to the updated key will only require it to be changed to `patch_step_size`, thus removing the suffixes for both 2D and 3D.
> (tagging @mariehbourget for confirmation - please feel free to edit if you want to add anything more. Thanks!)

Thanks @naga-karthik for the summary, looks good to me!

As a reference, here is how we handled modifying KEY names in the config file in the past with the config manager for retrocompatibility of old config files (e.g. config files of models used by SCT or ADS).
https://github.com/ivadomed/ivadomed/blob/ab381409e07af1f27a77e0cfb246f33bcc6291bb/ivadomed/config_manager.py#L73-L76

https://github.com/ivadomed/ivadomed/blob/ab381409e07af1f27a77e0cfb246f33bcc6291bb/ivadomed/config_manager.py#L119-L137

and

https://github.com/ivadomed/ivadomed/blob/ab381409e07af1f27a77e0cfb246f33bcc6291bb/ivadomed/config_manager.py#L147-L172
